### PR TITLE
Improvement on tests: MstTransportGrpc testable without server 

### PR DESCRIPTION
**Pull request by Exctues**
### Description of the Change
Now it is possible to tests MstTransportGrpc without running server.

### CODE REVIEW

File: test/module/irohad/multi_sig_transactions/transport_test.cpp [hunk 1]
```diff
@@ -252,4 +250,4 @@ TEST_F(TransportTest, ReplayAttack) {
 
   transport->SendState(&context, &proto_state, &response);
   transport->SendState(&context, &proto_state, &response);
-}
+}
```
Comment (lebdron) [inline test/module/irohad/multi_sig_transactions/transport_test.cpp]: Please add newline at the end of file. 
To automate, check the link https://stackoverflow.com/a/16761228

File: test/module/irohad/multi_sig_transactions/transport_test.cpp [hunk 2]
```diff
@@ -101,22 +148,8 @@ TEST_F(TransportTest, SendAndReceive) {
...
-                                         std::move(batch_factory_),
-                                         std::move(tx_presence_cache_),
-                                         completer_,
-                                         my_key_.publicKey(),
-                                         getTestLogger("MstState"),
-                                         getTestLogger("MstTransportGrpc"));
-  transport->subscribe(mst_notification_transport_);
-
   std::mutex mtx;
   std::condition_variable cv;
```
Comment (lebdron) [inline test/module/irohad/multi_sig_transactions/transport_test.cpp]: Mutex and condition variable can be removed, since there is no more multi-threading in the test.

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp [hunk 3]
```diff
@@ -14,20 +14,18 @@
...
 #include "logger/logger.hpp"
+#include "mst_mock.grpc.pb.h"
 #include "validators/field_validator.hpp"
 
 using namespace iroha;
 using namespace iroha::network;
 
 using iroha::ConstRefState;
 
-void sendStateAsyncImpl(const shared_model::interface::Peer &to,
```
[the remark this answers is not in the document]
Comment (Exctues) [reply irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp]: Didn't know about this codestyle moment. The declaration is returned back in the last fix.

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp [hunk 4]
```diff
@@ -147,28 +145,20 @@ void MstTransportGrpc::subscribe(
   subscriber_ = notification;
 }
 
-void MstTransportGrpc::sendState(const shared_model::interface::Peer &to,
```
[the remark this answers is not in the document]
Comment (Exctues) [reply irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp]: Done. The last fix makes minimum changes in files overall.

File: test/module/irohad/multi_sig_transactions/transport_test.cpp (hunk 1 above)
Comment (Exctues) [reply test/module/irohad/multi_sig_transactions/transport_test.cpp]: Thanks for the link! It is fixed now.

File: test/module/irohad/multi_sig_transactions/transport_test.cpp (hunk 2 above)
Comment (Exctues) [reply test/module/irohad/multi_sig_transactions/transport_test.cpp]: You are right, they are redundant. Already removed.

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp [hunk 5]
```diff
@@ -14,20 +14,18 @@
 #include "interfaces/iroha_internal/transaction_batch.hpp"
 #include "interfaces/transaction.hpp"
 #include "logger/logger.hpp"
+#include "mst_mock.grpc.pb.h"
```
[the remark this answers is not in the document]
Comment (Exctues) [reply irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp]: Removed.

File: test/module/irohad/multi_sig_transactions/transport_test.cpp [hunk 6]
```diff
@@ -59,6 +101,20 @@ class TransportTest : public ::testing::Test {
...
+      shared_model::validation::MockValidator<iroha::protocol::Transaction>>
+      proto_tx_validator;
+  std::shared_ptr<shared_model::proto::ProtoTransportFactory<
+      shared_model::interface::Transaction,
+      shared_model::proto::Transaction>>
+      tx_factory;
+  std::shared_ptr<MstTransportGrpc> transport;
+  std::shared_ptr<shared_model::interface::Peer> peer;
+  // stub will be deleted by unique_ptr created in client_creator
+  iroha::network::transport::MockMstTransportGrpcStub *stub;
```
Comment (MBoldyrev) [inline test/module/irohad/multi_sig_transactions/transport_test.cpp]: Again, codestyle. Here and forth
```suggestion
  iroha::network::transport::MockMstTransportGrpcStub *stub_;
```
https://google.github.io/styleguide/cppguide.html#Variable_Names

File: test/module/irohad/multi_sig_transactions/transport_test.cpp [hunk 7]
```diff
@@ -27,26 +29,66 @@ using namespace shared_model::interface;
...
+      : my_key_(makeKey()),
+        stub(new iroha::network::transport::MockMstTransportGrpcStub()) {}
+  void SetUp() override {
+    async_call_ = std::make_shared<AsyncGrpcClient<google::protobuf::Empty>>(
+        getTestLogger("AsyncClient"));
+    parser_ = std::make_shared<TransactionBatchParserImpl>();
+    batch_validator_ =
+        std::make_shared<shared_model::validation::BatchValidator>(
+            iroha::test::kTestsValidatorsConfig);
+    batch_factory_ =
```
Comment (MBoldyrev) [inline test/module/irohad/multi_sig_transactions/transport_test.cpp]: just wondering, why did you move the initialization of some members from constructor to `SetUp()`?

File: test/module/irohad/multi_sig_transactions/transport_test.cpp [hunk 8]
```diff
@@ -77,16 +133,6 @@ static bool statesEqual(const iroha::MstState &a, const iroha::MstState &b) {
...
-      std::make_unique<shared_model::validation::MockValidator<
-          shared_model::interface::Transaction>>();
-  auto proto_tx_validator = std::make_unique<
-      shared_model::validation::MockValidator<iroha::protocol::Transaction>>();
-  auto tx_factory = std::make_shared<shared_model::proto::ProtoTransportFactory<
-      shared_model::interface::Transaction,
-      shared_model::proto::Transaction>>(std::move(interface_tx_validator),
-                                         std::move(proto_tx_validator));
-
   ON_CALL(*tx_presence_cache_,
```
Comment (MBoldyrev) [inline test/module/irohad/multi_sig_transactions/transport_test.cpp]: Not your code, though this will show a warning, so better change it to
```suggestion
   EXPECT_CALL(*tx_presence_cache_,
```
and `WillByDefault` to `WillRepeatedly`

File: test/module/irohad/multi_sig_transactions/transport_test.cpp (hunk 7 above)
Comment (Exctues) [reply test/module/irohad/multi_sig_transactions/transport_test.cpp]: This change may protect someone who will unintentionally add side effect while writing new tests. This won't make the test run much longer but can save someone's time during test debugging.

File: test/module/irohad/multi_sig_transactions/transport_test.cpp (hunk 6 above)
Comment (lebdron) [reply test/module/irohad/multi_sig_transactions/transport_test.cpp]: This field is public.

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp [hunk 9]
```diff
@@ -150,26 +155,33 @@ void MstTransportGrpc::subscribe(
...
-          to.address(), grpc::InsecureChannelCredentials()));
-
+void sendStateAsyncImpl(
+    const shared_model::interface::Peer &to,
+    ConstRefState state,
+    const std::string &sender_key,
+    AsyncGrpcClient<google::protobuf::Empty> &async_call,
+    std::function<std::unique_ptr<transport::MstTransportGrpc::StubInterface>(
+        const shared_model::interface::Peer &)> client_creator) {
+  auto client = (client_creator == nullptr)
```
[the remark this answers is not in the document]
Comment (lebdron) [reply irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp]: As I understand, `nullptr` is used to reduce the changes in the pull request. In general this should not be neither boost::none, nor nullptr, but it would require to refactor the test. I suggest to add a TODO to fix the tests instead.

File: test/module/irohad/multi_sig_transactions/transport_test.cpp (hunk 8 above)
Comment (Exctues) [reply test/module/irohad/multi_sig_transactions/transport_test.cpp]: Changed.

File: irohad/multi_sig_transactions/transport/mst_transport_grpc.hpp [hunk 10]
```diff
@@ -88,6 +91,9 @@ namespace iroha {
       logger::LoggerPtr mst_state_logger_;  ///< Logger for created MstState
                                             ///< objects.
       logger::LoggerPtr log_;               ///< Logger for local use.
+      std::function<std::unique_ptr<transport::MstTransportGrpc::StubInterface>(
+          const shared_model::interface::Peer &)>
+          client_creator_;
```
[the remark this answers is not in the document]
Comment (Exctues) [reply irohad/multi_sig_transactions/transport/mst_transport_grpc.hpp]: Changed to sender_factory_

File: irohad/multi_sig_transactions/transport/mst_transport_grpc.hpp [hunk 11]
```diff
@@ -45,7 +45,10 @@ namespace iroha {
           std::shared_ptr<Completer> mst_completer,
           shared_model::crypto::PublicKey my_key,
           logger::LoggerPtr mst_state_logger,
-          logger::LoggerPtr log);
+          logger::LoggerPtr log,
+          std::function<
+              std::unique_ptr<transport::MstTransportGrpc::StubInterface>(
```
[the remark this answers is not in the document]
Comment (Exctues) [reply irohad/multi_sig_transactions/transport/mst_transport_grpc.hpp]: Aliased as SenderFactory

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (Exctues) [reply irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp]: I have changed this to boost::optional so that there is no ternary operator in client declaration.

File: test/module/irohad/multi_sig_transactions/transport_test.cpp (hunk 7 above)
Comment (Exctues) [reply test/module/irohad/multi_sig_transactions/transport_test.cpp]: This change also helps you to reduce the amount of code since you don't need to duplicate the declaration of variables anymore. That is because SetUp() called automatically per each test.
